File: skills/academic-figure/templates/rf-0044-clustered-network/r/plot.R
#!/usr/bin/env Rscript

fail <- function(msg) { message("ERROR: ", msg); quit(status = 2) }
args <- commandArgs(trailingOnly = TRUE)
opts <- list(nodes = NULL, edges = NULL, `output-prefix` = NULL, title = "Supplied network structure", dpi = "320")
for (arg in args) {
  if (!startsWith(arg, "--") || !grepl("=", arg, fixed = TRUE)) fail(paste0("arguments must use --name=value: ", arg))
  p <- strsplit(sub("^--", "", arg), "=", fixed = TRUE)[[1]]; key <- p[1]; value <- paste(p[-1], collapse = "=")
  if (!key %in% names(opts)) fail(paste0("unknown argument --", key)); opts[[key]] <- value
}
if (any(vapply(opts[c("nodes", "edges", "output-prefix")], is.null, logical(1)))) fail("--nodes, --edges, and --output-prefix are required")
dpi <- suppressWarnings(as.integer(opts$dpi)); if (is.na(dpi) || dpi < 72) fail("--dpi must be at least 72")
n <- tryCatch(read.csv(opts$nodes, stringsAsFactors = FALSE, check.names = FALSE), error = function(e) fail(conditionMessage(e)))
e <- tryCatch(read.csv(opts$edges, stringsAsFactors = FALSE, check.names = FALSE), error = function(e) fail(conditionMessage(e)))
nreq <- c("node_id", "label", "x", "y", "cluster"); ereq <- c("source", "target")
if (!all(nreq %in% names(n)) || nrow(n) == 0) fail(paste("nodes CSV must contain", paste(nreq, collapse = ", ")))
if (!all(ereq %in% names(e)) || nrow(e) == 0) fail(paste("edges CSV must contain", paste(ereq, collapse = ", ")))
if (nrow(n) < 5 || nrow(n) > 120) fail("nodes CSV must contain 5–120 rows")
if (nrow(e) > 500) fail("at most 500 edges are supported")
for (nm in c("node_id", "label", "cluster")) n[[nm]] <- trimws(as.character(n[[nm]]))
if (any(n$node_id == "") || anyDuplicated(n$node_id) || any(n$label == "") || any(n$cluster == "")) fail("node_id must be unique and label/cluster non-empty")
n$x <- suppressWarnings(as.numeric(n$x)); n$y <- suppressWarnings(as.numeric(n$y))
if (!"size" %in% names(n)) n$size <- 1; n$size <- suppressWarnings(as.numeric(n$size))
if (any(!is.finite(n$x)) || any(!is.finite(n$y)) || any(!is.finite(n$size)) || any(n$size <= 0)) fail("coordinates must be finite and size positive")
if (!"show_label" %in% names(n)) n$show_label <- FALSE
raw_show <- tolower(trimws(as.character(n$show_label)))
if (any(!raw_show %in% c("true", "false", "1", "0", "yes", "no", ""))) fail("show_label must be true or false")
n$show_label <- raw_show %in% c("true", "1", "yes")
clusters <- unique(n$cluster); if (length(clusters) > 12) fail("at most 12 clusters are supported")
for (nm in c("source", "target")) e[[nm]] <- trimws(as.character(e[[nm]]))
if (any(!e$source %in% n$node_id) || any(!e$target %in% n$node_id) || any(e$source == e$target)) fail("edge endpoints must reference two different known nodes")
edge_key <- apply(cbind(e$source, e$target), 1, function(z) paste(sort(z), collapse = "\r"))
if (anyDuplicated(edge_key)) fail("duplicate undirected edges are not allowed")
if (!"weight" %in% names(e)) e$weight <- 1; e$weight <- suppressWarnings(as.numeric(e$weight))
if (any(!is.finite(e$weight)) || any(e$weight <= 0)) fail("edge weight must be finite and positive")
if (!"edge_group" %in% names(e)) e$edge_group <- "edge"; e$edge_group <- trimws(as.character(e$edge_group)); e$edge_group[e$edge_group == ""] <- "edge"
groups <- unique(e$edge_group)
scale_to <- function(x, lo, hi) if (max(x) == min(x)) rep((lo + hi) / 2, length(x)) else lo + (x - min(x)) / (max(x) - min(x)) * (hi - lo)
palette <- c("#C8755C", "#4E8995", "#887DA5", "#789566", "#BE974A", "#A96775")
cols <- setNames(rep(palette, length.out = length(clusters)), clusters)
line_types <- setNames(seq_along(groups), groups); line_types[line_types > 2] <- 3
node_cex <- scale_to(n$size, 1.5, 2.8); edge_lwd <- scale_to(e$weight, 0.8, 2.3)
draw <- function(device_fun) {
  device_fun(); par(bg = "#FBFAF7", mar = c(1.2, 1.2, 5, 1.2), family = "sans", xpd = NA)
  xr <- range(n$x); yr <- range(n$y); padx <- max(diff(xr), 1) * 0.18; pady <- max(diff(yr), 1) * 0.18
  plot.new(); plot.window(xlim = xr + c(-padx, padx), ylim = yr + c(-pady, pady), asp = 1)
  for (i in seq_len(nrow(e))) {
    a <- match(e$source[i], n$node_id); b <- match(e$target[i], n$node_id)
    segments(n$x[a], n$y[a], n$x[b], n$y[b], col = adjustcolor("#91A0A5", alpha.f = 0.5), lwd = edge_lwd[i], lty = line_types[[e$edge_group[i]]])
  }
  points(n$x, n$y, pch = 21, bg = cols[n$cluster], col = "white", lwd = 1.4, cex = node_cex)
  if (any(n$show_label)) text(n$x[n$show_label], n$y[n$show_label], n$label[n$show_label], pos = 3, offset = 0.9, cex = 0.76, font = 2, col = "#20272D")
  legend("topleft", legend = clusters, pch = 21, pt.bg = cols[clusters], pt.cex = 1.15, col = "white", bty = "n", horiz = length(clusters) <= 4, cex = 0.76, inset = 0.01)
  title(main = opts$title, adj = 0, line = 2.25, cex.main = 1.35, col.main = "#20272D", font.main = 2)
  mtext("Coordinates, clusters, labels and edge weights are supplied; no network inference is performed.", side = 3, adj = 0, line = 0.55, cex = 0.74, col = "#68767D")
  dev.off()
}
prefix <- opts$`output-prefix`; dir.create(dirname(prefix), recursive = TRUE, showWarnings = FALSE)
draw(function() png(paste0(prefix, ".png"), width = 9.2, height = 7, units = "in", res = dpi, bg = "#FBFAF7"))
draw(function() pdf(paste0(prefix, ".pdf"), width = 9.2, height = 7, bg = "#FBFAF7", useDingbats = FALSE))
message(sprintf("Validated %d nodes, %d unique undirected edges, and %d supplied clusters; excluded rows: 0.", nrow(n), nrow(e), length(clusters)))

File: skills/academic-figure/templates/rf-0044-clustered-network/python/plot.py
#!/usr/bin/env python3
"""Render a validated network with supplied coordinates and clusters."""
from __future__ import annotations

import argparse
import csv
import math
import sys
from pathlib import Path

import matplotlib.pyplot as plt
from matplotlib.lines import Line2D

INK = "#20272D"
MUTED = "#68767D"
BG = "#FBFAF7"
PALETTE = ["#C8755C", "#4E8995", "#887DA5", "#789566", "#BE974A", "#A96775"]


def fail(msg: str) -> None:
    print(f"ERROR: {msg}", file=sys.stderr)
    raise SystemExit(2)


def args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--nodes", required=True)
    p.add_argument("--edges", required=True)
    p.add_argument("--output-prefix", required=True)
    p.add_argument("--title", default="Supplied network structure")
    p.add_argument("--dpi", type=int, default=320)
    return p.parse_args()


def read_csv(path: Path):
    with path.open(newline="", encoding="utf-8-sig") as f:
        return list(csv.DictReader(f))


def boolean(raw: str, row: int) -> bool:
    value = raw.strip().lower()
    if value in {"", "false", "0", "no"}:
        return False
    if value in {"true", "1", "yes"}:
        return True
    fail(f"nodes row {row}: show_label must be true or false")


def load(nodes_path: Path, edges_path: Path):
    nr, er = read_csv(nodes_path), read_csv(edges_path)
    nreq, ereq = {"node_id", "label", "x", "y", "cluster"}, {"source", "target"}
    if not nr or not nreq.issubset(nr[0]):
        fail(f"nodes CSV must contain {', '.join(sorted(nreq))}")
    if not er or not ereq.issubset(er[0]):
        fail(f"edges CSV must contain {', '.join(sorted(ereq))}")
    if not 5 <= len(nr) <= 120:
        fail("nodes CSV must contain 5–120 rows")
    if len(er) > 500:
        fail("at most 500 edges are supported")
    nodes = {}
    clusters = []
    for i, row in enumerate(nr, 2):
        nid, label, cluster = [(row.get(k) or "").strip() for k in ("node_id", "label", "cluster")]
        if not nid or nid in nodes or not label or not cluster:
            fail(f"nodes row {i}: node_id must be unique and label/cluster non-empty")
        try:
            x, y = float(row["x"]), float(row["y"])
            size = float((row.get("size") or "1").strip())
        except ValueError:
            fail(f"nodes row {i}: x, y, and size must be numeric")
        if not all(math.isfinite(v) for v in (x, y, size)) or size <= 0:
            fail(f"nodes row {i}: coordinates must be finite and size positive")
        nodes[nid] = {"id": nid, "label": label, "x": x, "y": y, "cluster": cluster,
                      "size": size, "show": boolean(row.get("show_label") or "", i)}
        if cluster not in clusters:
            clusters.append(cluster)
    if len(clusters) > 12:
        fail("at most 12 clusters are supported")
    edges, seen = [], set()
    groups = []
    for i, row in enumerate(er, 2):
        source, target = [(row.get(k) or "").strip() for k in ("source", "target")]
        if source not in nodes or target not in nodes or source == target:
            fail(f"edges row {i}: endpoints must reference two different known nodes")
        key = tuple(sorted((source, target)))
        if key in seen:
            fail(f"edges row {i}: duplicate undirected edge {key[0]}–{key[1]}")
        seen.add(key)
        try:
            weight = float((row.get("weight") or "1").strip())
        except ValueError:
            fail(f"edges row {i}: weight must be numeric")
        if not math.isfinite(weight) or weight <= 0:
            fail(f"edges row {i}: weight must be finite and positive")
        group = (row.get("edge_group") or "edge").strip() or "edge"
        if group not in groups:
            groups.append(group)
        edges.append({"source": source, "target": target, "weight": weight, "group": group})
    return nodes, edges, clusters, groups


def scale(values, low, high):
    lo, hi = min(values), max(values)
    if hi == lo:
        return [0.5 * (low + high)] * len(values)
    return [low + (v - lo) / (hi - lo) * (high - low) for v in values]


def render(nodes, edges, clusters, groups, title: str, out: Path, dpi: int):
    colors = {g: PALETTE[i % len(PALETTE)] for i, g in enumerate(clusters)}
    line_styles = {g: ("-" if i == 0 else (0, (3, 3))) for i, g in enumerate(groups)}
    widths = scale([e["weight"] for e in edges], 0.7, 2.2)
    fig, ax = plt.subplots(figsize=(9.2, 7.0), facecolor=BG)
    ax.set_facecolor(BG)
    for edge, width in zip(edges, widths):
        a, b = nodes[edge["source"]], nodes[edge["target"]]
        ax.plot([a["x"], b["x"]], [a["y"], b["y"]], color="#91A0A5", alpha=0.42,
                linewidth=width, linestyle=line_styles[edge["group"]], zorder=1)
    sizes = scale([n["size"] for n in nodes.values()], 160, 520)
    for (node, area) in zip(nodes.values(), sizes):
        ax.scatter(node["x"], node["y"], s=area, color=colors[node["cluster"]],
                   edgecolor="white", linewidth=1.4, zorder=3)
        if node["show"]:
            ax.annotate(node["label"], (node["x"], node["y"]), xytext=(0, 12),
                        textcoords="offset points", ha="center", va="bottom", fontsize=8.5,
                        color=INK, fontweight="semibold", zorder=4)
    handles = [Line2D([0], [0], marker="o", color="none", markerfacecolor=colors[g],
                      markeredgecolor="white", markersize=9, label=g) for g in clusters]
    ax.legend(handles=handles, loc="upper left", bbox_to_anchor=(0.01, 0.99), frameon=False,
              ncol=min(3, len(handles)), fontsize=8.5, handletextpad=0.4, columnspacing=1.2)
    fig.suptitle(title, x=0.075, y=0.965, ha="left", fontsize=18, fontweight="semibold", color=INK)
    fig.text(0.075, 0.925, "Coordinates, clusters, labels and edge weights are supplied; no network inference is performed.",
             fontsize=9, color=MUTED)
    xs, ys = [n["x"] for n in nodes.values()], [n["y"] for n in nodes.values()]
    padx, pady = max(max(xs)-min(xs), 1)*0.18, max(max(ys)-min(ys), 1)*0.18
    ax.set_xlim(min(xs)-padx, max(xs)+padx); ax.set_ylim(min(ys)-pady, max(ys)+pady)
    ax.set_aspect("equal", adjustable="box"); ax.axis("off")
    fig.subplots_adjust(left=0.06, right=0.96, top=0.88, bottom=0.05)
    out.parent.mkdir(parents=True, exist_ok=True)
    fig.savefig(out.with_suffix(".png"), dpi=dpi, facecolor=BG)
    fig.savefig(out.with_suffix(".svg"), facecolor=BG)
    plt.close(fig)


def main():
    a = args()
    nodes, edges, clusters, groups = load(Path(a.nodes), Path(a.edges))
    render(nodes, edges, clusters, groups, a.title, Path(a.output_prefix), a.dpi)
    print(f"Validated {len(nodes)} nodes, {len(edges)} unique undirected edges, and {len(clusters)} supplied clusters; excluded rows: 0.")


if __name__ == "__main__":
    main()
